docs/overview/whatisvpp/softwarearchitecture.rst

```diff
@@ -0,0 +1,707 @@
+.. _softwarearchitecture:
+
+.. toctree::
+
+Software Architecture
+=====================
+
+The fd.io vpp implementation is a third-generation vector packet
+processing implementation specifically related to US Patent 7,961,636,
+as well as earlier work. Note that the Apache-2 license specifically
+grants non-exclusive patent licenses; we mention this patent as a
+point of historical interest.
+
+For performance, the vpp dataplane consists of a directed graph of
+forwarding nodes which process multiple packets per invocation. This
+schema enables a variety of cache-related optimizations: pipelining
+and prefetching to cover dependent read latency, inherent I-cache
+phase behavior. Aside from hardware input and hardware output nodes,
+the entire forwarding graph is portable code.
+
+Depending on the scenario at hand, we often spin up multiple worker
+threads which process ingress-hashes packets from multiple queues using
+identical forwarding graph replicas. 
+
+Implemetation taxonomy
+----------------------
+
+The vpp dataplane consists of four distinct layers: 
+
+* An infrastructure layer comprising vppinfra, vlib, svm, and binary api libraries. See .../src/{vppinfra, vlib, svm, vlibapi, vlibmemory}
+
+* A generic network stack layer: vnet. See .../src/vnet
+
+* An application shell: vpp. See .../src/vpp
+
+* An increasingly rich set of data-plane plugins: see .../src/plugins
+
+It's important to understand each of these layers in a certain amount
+of detail. Much of the implementation is best dealt with at the API
+level and otherwise left alone.
+
+Vppinfra
+========
```

Comment:
At the moment Vppinfra, Vlib etc are not appearing as  a subheadings of "Software Architecture" in the TOC, they appear at the same heading level. In Restructured Text, document structure is determined from the succession of heading characters, i.e. the order of heading character determine heading level. 

So to solve this problem you need to insert Overview subsection above and to assign it a different character. 
Python uses the following convention. 
https://devguide.python.org/documenting/#sections

    # with overline, for parts
    * with overline, for chapters
    =, for sections
    -, for subsections
    ^, for subsubsections
    ", for paragraphs